skyportal/handlers/api/internal/source_views.py

```diff
@@ -0,0 +1,50 @@
+import datetime
+from sqlalchemy import func, desc
+from baselayer.app.access import permissions, auth_or_token
+from ...base import BaseHandler
+from ....models import (
+    DBSession, Source, SourceView, GroupSource
+)
+
+
+class SourceViewsHandler(BaseHandler):
+    @auth_or_token
+    def get(self):
+        prefs = (self.current_user.preferences if
+                 hasattr(self.current_user, 'preferences') and
+                 self.current_user.preferences else {})
+        if 'topSources' in prefs and 'maxNumSources' in prefs['topSources']:
+            max_num_sources = int(prefs['topSources']['maxNumSources'])
+        else:
+            max_num_sources = 10
+        if 'topSources' in prefs and 'sinceDaysAgo' in prefs['topSources']:
+            since_days_ago = int(prefs['topSources']['sinceDaysAgo'])
+        else:
+            since_days_ago = 30
+        cutoff_day = datetime.datetime.now() - datetime.timedelta(days=since_days_ago)
+        q = (DBSession.query(func.count(SourceView.source_id).label('views'),
+                             SourceView.source_id).group_by(SourceView.source_id)
+             .filter(SourceView.source_id.in_(DBSession.query(
+                 GroupSource.source_id).filter(GroupSource.group_id.in_(
+                     [g.id for g in self.current_user.groups]))))
+             .filter(SourceView.created_at >= cutoff_day)
+             .order_by(desc('views')).limit(max_num_sources))
+        return self.success(data={'sourceViews': q.all()})
+
+    @auth_or_token
+    def post(self, source_id):
+        # Ensure token/user has access to source
+        s = Source.get_if_owned_by(source_id, self.current_user)
+        # This endpoint will only be hit by front-end, so this will never be a token
+        register_source_view(source_id=source_id,
```

Comment:
Why does post get a source view?

Reply:
This is the post handler of the `SourceViewHandler`, which is what we hit to register a front-end view (we don't just tally ever front-end GET request, because that happens too often, e.g. when comments are added, the source is re-fetched. So we handle the logic of when to log a source view on the front-end, and from there we hit `/api/internal/source_views` with a POST, which is what this handles)

Reply:
Aha, thanks!